[WIP] [jslc] Port from antlr3 to antlr4.

brcolow opened the pull request:
Remove all embedded actions from the grammar file and use antlr4's new visitor mechanism.

antlr3 has not had any updates since 2014 and antlr4 was (according to the antlr author) a complete re-write. One of the biggest changes in this PR is that all embedded actions are removed from the antlr grammar file and instead they are implemented using the antlr4 visitor mechanism. Beyond just being a cleaner approach to making antlr parsers it allows us to more easily extend the grammar (and add support for higher level shader language features) down the line, which I intend to do as part of my much larger work I am doing with adding a DirectX 11 pipeline and run-time compiled, user-specified, JSL language shaders that can be applied to the JavaFX scene graph.

Seen in isolation this may PR may look to be an unnecessary risk but I believe that it will (beyond keeping the libraries that JavaFX uses up-to-date) more easily enable future innovation/experimentation.

I found while working on JSLC that the tests are not run as part of the gradle build (no matter the value of any various gradle properties). In order to make sure that the new tests pass I made a stand-alone sandbox project. Ideally I believe that the tests should be run (perhaps behind a JSL_TEST gradle property) as part of the build but I am not sure how to add such a test task (it is beyond my gradle-fu). Having the tests run as part of the build would reduce the risk of merging this PR.

I decided to separate out this independent work from my larger work I am doing because I figured it would make it easier to review but don't want to give the impression that I think it's a good stand-alone PR, it is part of a larger goal.

If the reception to this PR is warm I will happily submit the requisite bug reports and RFE.

kevinrushforth commented:
I'll need to do an initial evaluation of this to make sure I understand the implications. It won't be right away, but I can get back to you on this.

Also, as with any third-party software we need legal approval for the update.

brcolow commented:
Sounds good, Kevin. I will continue to work on the other, cooler things that will use this as a base. Does getting legal approval require me to submit anything? Or is it an in-house process? antlr4 was written by same guy that did antlr3 (Dr. Terence Parr) and it uses the same license (BSD license).

kevinrushforth commented:
> Does getting legal approval require me to submit anything? Or is it an in-house process?

It's an in-house process.

> antlr4 was written by same guy that did antlr3 (Dr. Terence Parr) and it uses the same license (BSD license).

Thanks.

brcolow commented:
I am going to run a diff over all of the shader files (and software versions) that JSLC generates and compare before this PR and after, as they should be identical. That will almost eliminate the risk of merging this PR.

kevinrushforth commented:
I have a few general questions:

1. Is it feasible to first upgrade to antlr4 without using any new antlr4 features, such that there are no changes to the grammar files and either no or minimal changes to JSLC itself? If so, then I would prefer to see this PR split into two separate issues (reviewing them together might make sense).

2. Is there a reason you switched from using antlr-complete (the equivalent for v4 is the `antlr4` artifact) to the three separate artifacts? I don't remember the details, but we switched to using the "complete" artifact in FX 9, so I was wondering if there was an advantage to using the separate ones?

3. Was renaming JSL.g to JSL.g4 necessary? If it follows the recommended convention for antlr4 then it is OK, but otherwise it seems unnecessary.

brcolow commented:
1.) I can definitely look at doing that. I will create a new branch that
upgrades to antlr4 and makes the absolute minimum
amount of changes and we can see what that looks like. Good idea.

2.) The only reason I switched is because I couldn't find antlr4-complete
on Maven Central listed as a separate artifact. But now
that I look more closely I see that the "-complete" artifact does appear to
be in the repository and so will make that change for sure.

3.) Renaming from g to g4 is I think necessary as the antlr4 docs specify "
The file name containing grammar X must be called X.g4".
I didn't actually try compiling with a ".g" extension though, so it may
work.

I will work on making two separate PRs and the changes to the complete
artifact.

Thanks a lot Kevin.

On Thu, Jan 31, 2019 at 12:39 PM Kevin Rushforth <notifications@github.com>
wrote:

> I have a few general questions:
>
>    1.
>
>    Is it feasible to first upgrade to antlr4 without using any new antlr4
>    features, such that there are no changes to the grammar files and either no
>    or minimal changes to JSLC itself? If so, then I would prefer to see this
>    PR split into two separate issues (reviewing them together might make
>    sense).
>    2.
>
>    Is there a reason you switched from using antlr-complete (the
>    equivalent for v4 is the antlr4 artifact) to the three separate
>    artifacts? I don't remember the details, but we switched to using the
>    "complete" artifact in FX 9, so I was wondering if there was an advantage
>    to using the separate ones?
>    3.
>
>    Was renaming JSL.g to JSL.g4 necessary? If it follows the recommended
>    convention for antlr4 then it is OK, but otherwise it seems unnecessary.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/javafxports/openjdk-jfx/pull/360#issuecomment-459477931>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGSNqukuD62iIjK_wV9pkeTStW-trJDiks5vI0ZmgaJpZM4aTu9E>
> .
>


-- 
Michael Ennen

brcolow commented:
1.) I can definitely look at doing that. I will create a new branch that upgrades to antlr4 and makes the absolute minimum amount of changes and we can see what that looks like. Good idea.

2.) The only reason I switched is because I couldn't find antlr4-complete on Maven Central listed as a separate artifact. But now that I look more closely I see that the "-complete" artifact does appear to be in the repository and so will make that change for sure.

3.) Renaming from g to g4 is I think necessary as the antlr4 docs specify " The file name containing grammar X must be called X.g4". I didn't actually try compiling with a ".g" extension though, so it may work. I will work on making two separate PRs and the changes to the complete artifact.

Thanks!

johanvos commented:
This starts looking good. I don't think this is something we should rush in 12, but we can do it early in 13. Do you create a JBS issue as well?

brcolow commented:
I have now created a JBS bug report for the first change set broken out to have just the minimum changes necessary to bump the antlr version: https://bugs.openjdk.java.net/browse/JDK-8218764

brcolow commented:
I will be creating more change sets that have follow-up PRs that fill in the functionality in this PR (and more). I will open them when https://github.com/javafxports/openjdk-jfx/pull/372 is merged - I wish Github handled dependent PRs better.

brcolow commented:
This was broken out into #372 and #413
